File: fe2-android/app/src/test/unit/java/com/github/dedis/popstellar/model/network/method/message/data/socialmedia/NotifyAddChirpTest.java
package com.github.dedis.popstellar.model.network.method.message.data.socialmedia;

import androidx.test.ext.junit.runners.AndroidJUnit4;

import com.github.dedis.popstellar.model.network.JsonUtilsTest;
import com.github.dedis.popstellar.model.network.method.message.data.Action;
import com.github.dedis.popstellar.model.network.method.message.data.Objects;
import com.github.dedis.popstellar.model.objects.Channel;
import com.github.dedis.popstellar.model.objects.security.MessageID;
import com.google.gson.JsonParseException;

import org.junit.Test;
import org.junit.runner.RunWith;

import static com.github.dedis.popstellar.testutils.Base64DataUtils.generateMessageID;
import static com.github.dedis.popstellar.testutils.Base64DataUtils.generateMessageIDOtherThan;
import static org.junit.Assert.*;

@RunWith(AndroidJUnit4.class)
public class NotifyAddChirpTest {

  private static final MessageID CHIRP_ID = generateMessageID();
  private static final Channel CHANNEL = Channel.fromString("/root/laoId/social/myChannel");
  private static final long TIMESTAMP = 1631280815;

  private static final NotifyAddChirp NOTIFY_ADD_CHIRP =
      new NotifyAddChirp(CHIRP_ID, CHANNEL, TIMESTAMP);

  @Test
  public void getObjectTest() {
    assertEquals(Objects.CHIRP.getObject(), NOTIFY_ADD_CHIRP.getObject());
  }

  @Test
  public void getActionTest() {
    assertEquals(Action.NOTIFY_ADD.getAction(), NOTIFY_ADD_CHIRP.getAction());
  }

  @Test
  public void getChirpIdTest() {
    assertEquals(CHIRP_ID, NOTIFY_ADD_CHIRP.getChirpId());
  }

  @Test
  public void getChannelTest() {
    assertEquals(CHANNEL, NOTIFY_ADD_CHIRP.getChannel());
  }

  @Test
  public void getTimestampTest() {
    assertEquals(TIMESTAMP, NOTIFY_ADD_CHIRP.getTimestamp());
  }

  @Test
  public void equalsTest() {
    assertEquals(NOTIFY_ADD_CHIRP, new NotifyAddChirp(CHIRP_ID, CHANNEL, TIMESTAMP));

    String random = "random";
    assertNotEquals(
        NOTIFY_ADD_CHIRP,
        new NotifyAddChirp(generateMessageIDOtherThan(CHIRP_ID), CHANNEL, TIMESTAMP));
    assertNotEquals(
        NOTIFY_ADD_CHIRP, new NotifyAddChirp(CHIRP_ID, Channel.ROOT.subChannel(random), TIMESTAMP));
    assertNotEquals(NOTIFY_ADD_CHIRP, new NotifyAddChirp(CHIRP_ID, CHANNEL, TIMESTAMP + 1));
  }

  @Test
  public void jsonValidationTest() {
    JsonUtilsTest.testData(NOTIFY_ADD_CHIRP);

    String pathDir = "protocol/examples/messageData/chirp_notify_add/";
    String jsonInvalid1 =
        JsonUtilsTest.loadFile(pathDir + "wrong_chirp_notify_add_negative_time.json");
    String jsonInvalid2 =
        JsonUtilsTest.loadFile(pathDir + "wrong_chirp_notify_add_not_base_64_chirp_id.json");
    assertThrows(JsonParseException.class, () -> JsonUtilsTest.parse(jsonInvalid1));
    assertThrows(JsonParseException.class, () -> JsonUtilsTest.parse(jsonInvalid2));
  }
}

File: fe2-android/app/src/test/unit/java/com/github/dedis/popstellar/model/network/method/message/data/socialmedia/NotifyAddChirpTest.kt
package com.github.dedis.popstellar.model.network.method.message.data.socialmedia

import androidx.test.ext.junit.runners.AndroidJUnit4
import com.github.dedis.popstellar.model.network.JsonTestUtils.loadFile
import com.github.dedis.popstellar.model.network.JsonTestUtils.parse
import com.github.dedis.popstellar.model.network.JsonTestUtils.testData
import com.github.dedis.popstellar.model.network.method.message.data.Action
import com.github.dedis.popstellar.model.network.method.message.data.Objects
import com.github.dedis.popstellar.model.objects.Channel
import com.github.dedis.popstellar.model.objects.Channel.Companion.fromString
import com.github.dedis.popstellar.testutils.Base64DataUtils
import com.google.gson.JsonParseException
import org.junit.Assert
import org.junit.Test
import org.junit.runner.RunWith

@RunWith(AndroidJUnit4::class)
class NotifyAddChirpTest {
  @Test
  fun objectTest() {
    Assert.assertEquals(Objects.CHIRP.`object`, NOTIFY_ADD_CHIRP.`object`)
  }

  @Test
  fun actionTest() {
    Assert.assertEquals(Action.NOTIFY_ADD.action, NOTIFY_ADD_CHIRP.action)
  }

  @Test
  fun chirpIdTest() {
    Assert.assertEquals(CHIRP_ID, NOTIFY_ADD_CHIRP.chirpId)
  }

  @Test
  fun channelTest() {
    Assert.assertEquals(CHANNEL, NOTIFY_ADD_CHIRP.channel)
  }

  @Test
  fun timestampTest() {
    Assert.assertEquals(TIMESTAMP, NOTIFY_ADD_CHIRP.timestamp)
  }

  @Test
  fun equalsTest() {
    Assert.assertEquals(NOTIFY_ADD_CHIRP, NotifyAddChirp(CHIRP_ID, CHANNEL, TIMESTAMP))
    val random = "random"
    Assert.assertNotEquals(
      NOTIFY_ADD_CHIRP,
      NotifyAddChirp(Base64DataUtils.generateMessageIDOtherThan(CHIRP_ID), CHANNEL, TIMESTAMP)
    )
    Assert.assertNotEquals(
      NOTIFY_ADD_CHIRP,
      NotifyAddChirp(CHIRP_ID, Channel.ROOT.subChannel(random), TIMESTAMP)
    )
    Assert.assertNotEquals(NOTIFY_ADD_CHIRP, NotifyAddChirp(CHIRP_ID, CHANNEL, TIMESTAMP + 1))
  }

  @Test
  fun jsonValidationTest() {
    testData(NOTIFY_ADD_CHIRP)
    val pathDir = "protocol/examples/messageData/chirp_notify_add/"
    val jsonInvalid1 = loadFile(pathDir + "wrong_chirp_notify_add_negative_time.json")
    val jsonInvalid2 = loadFile(pathDir + "wrong_chirp_notify_add_not_base_64_chirp_id.json")

    Assert.assertThrows(JsonParseException::class.java) { parse(jsonInvalid1) }
    Assert.assertThrows(JsonParseException::class.java) { parse(jsonInvalid2) }
  }

  companion object {
    private val CHIRP_ID = Base64DataUtils.generateMessageID()
    private val CHANNEL = fromString("/root/laoId/social/myChannel")
    private const val TIMESTAMP: Long = 1631280815
    private val NOTIFY_ADD_CHIRP = NotifyAddChirp(CHIRP_ID, CHANNEL, TIMESTAMP)
  }
}
